### PR TITLE
Mobile My Books Redesign #7376

### LINKED ISSUE #6857: My Books Page: mobile Redesign
Designs: https://www.figma.com/proto/KE0BoBAlMPzPHbhvVblKS8/My-Books-Page-Prototype-Flows?node-id=332%3A80&scaling=scale-down&page-id=277%3A2&starting-point-node-id=332%3A80&show-proto-sidebar=1

Implement designs from #6875

image

## Problem

Currently the My Books page (e.g. https://openlibrary.org/account/books) navigation has not been styled for accessibility on mobile. The desktop sidebar navigation, on mobile, takes up nearly half of the page. It becomes unclear to some patrons that it is a navigation menu v. the page's content:

## Possible Solution

We're exploring a redesign of the My Books page (mobile + navigation):
* https://www.figma.com/file/32L8oj6wNwydCTZVp4Vcfr/Open-Library?node-id=0%3A1
* @szgrune do you have an updated figma link to add here?

www figma com_file_32L8oj6wNwydCTZVp4Vcfr_Open-Library_node-id=0%3A1 (1)
www figma com_file_32L8oj6wNwydCTZVp4Vcfr_Open-Library_node-id=0%3A1 (2)
www figma com_file_32L8oj6wNwydCTZVp4Vcfr_Open-Library_node-id=0%3A1 (3)

## Subtasks
* [x] #6858
* [x] Test + choose a design ([Design](https://user-images.githubusercontent.com/978325/209074998-ae84170a-846a-4e6c-86c1-6e84f2e27fb4.png))
* [x] #7376

### Relevant url?

https://openlibrary.org/account/books

### Details

- **Logged in (Y/N)?**
- **Browser type/version?**
- **Operating system?**
- **Environment (prod/dev/local)?** prod

### LINKED ISSUE #7376: My Books: Mobile Re-design
Designs: https://www.figma.com/proto/KE0BoBAlMPzPHbhvVblKS8/My-Books-Page-Prototype-Flows?node-id=332%3A80&scaling=scale-down&page-id=277%3A2&starting-point-node-id=332%3A80&show-proto-sidebar=1

Implement designs from #6875
 
image

### PR DESCRIPTION
**Pull request by szgrune**
Closes #6857

Update to initial prototype of My Books mobile menu interface including functional book showcase component and View All Lists button, limited number of lists displayed

Reorganizes My Books menu on mobile, tweaks css to improve usability/accessibility, renders books in mobile 'showcase' carousel components, cleans up code, fixes issue of potentially infinite lists above the My Stats menu section

### Technical

This implementation is a first draft and still has major issues with the way the book covers are being rendered. My hope is to finish this next week with support from staff

Edit (January 20, 2023): the book cover issue seems to mostly be fixed (thanks to @cdrini for your help!). I also added an index variable to render_carousel_cover() parameters in mobile_carousel.html to enable adding a gradient mask-image for last book when five books are displayed. Also added a gap and changed padding for book covers in showcase carousel, changed background color to light grey

Edit (February 13, 2023): moved css from mobile_carousel.html into mybooks-details.less next to other relevant carousel header css; capped number of lists displayed on mobile My Books menu to 3 and added View All Lists button at bottom

### Testing

Use a responsive developer tool to view account/books on different sized mobile displays and test out the new menu

### Screenshot

Jan 19 version:
mybooks-mobile-draft

Jan 20 version:
mobile my books prototype screenshot

Feb 13 version (current):
my books mobile View All Lists

### Stakeholders

@mekarpeles @cdrini @jimchamp

### CODE REVIEW

**Comment by szgrune** [comment]
Breadcrumbs added! 

My Books mobile w breadcrumbs

**Comment by szgrune** [comment]
The most recent commit mostly closes #6857, though some more testing and optimization is certainly needed.

This  

Update to initial prototype of My Books mobile menu interface including new landing page menu, functional book showcase component, view all lists button, and breadcrumb dropdowns for each reading log or list page, as well as notes/reviews (new breadcrumbs currently on mobile and desktop). 

Stats is available as an option on these dropdown menus but doesn't include a dropdown menu of its own as it is a child with no siblings of the Already Read section. 

Dropdown menus generally show sibling pages, or in the case of individual List pages just include a link to the View All Lists page. In the future, I want to add more menu items to the individual List page dropdown buttons such as the user's 5 most recently created lists.

### Technical

This most recent commit added breadcrumb dropdowns to the individual reading log pages and the individual list pages, as well as cleaning up some issues with code style and organization. 

The breadcrumb dropdown was created as a new component in the file breadcrumb_select.html, requiring the name of the currently selected page as a string and a list of options (tuples with a string title and page URL for each menu option). The lists of options for different My Books pages are all hard coded in books.html

### Testing
Used responsive developer tool to view account/books and each of the reading log pages, as well as list and lists, to check for bugs in new My Books landing page and breadcrumb dropdown component for the child pages

### Screenshots

#### My Books landing page:
Screenshot 2023-04-04 at 7 24 43 PM

#### Reading log breadcrumb menu closed:
Screenshot 2023-04-04 at 7 22 37 PM

#### Reading log breadcrumb menu open:
Screenshot 2023-04-04 at 7 23 44 PM

#### Reading log breadcrumb on desktop:
Screenshot 2023-04-04 at 7 25 44 PM

#### List breadcrumb closed (mobile):
Screenshot 2023-04-04 at 7 26 28 PM

#### List breadcrumb open (mobile):
Screenshot 2023-04-04 at 7 26 44 PM

File: openlibrary/plugins/upstream/mybooks.py [hunk 1]
```diff
@@ -287,13 +287,13 @@ def get_shelf(name, page=1):
             # Ideally, do all 3 lookups in one add_availability call
             want_to_read.docs = add_availability(
                 [d for d in want_to_read.docs if d.get('title')]
-            )[:4]
+            )[:5]
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/mybooks.py:290]: the value [5] may conflict with what's needed on desktop -- we may need a special solution on the frontend for desktop to hide the 5th item so there isn't a scrollbar for each carousel.

File: openlibrary/templates/books/breadcrumb_select.html [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            border-radius: 4px;
+        }
+
+        .down-chevron {
+            margin: 6px 10px 0px 6px;
+            display: inline-block;
+            transform: rotate(90deg);
+            z-index: 1;
+        }
+    </style>
```
Comment (mekarpeles) [inline openlibrary/templates/books/breadcrumb_select.html:50]: Can these be moved to breadcrumbs.less? Or existing mybooksdetails.less

Comment (szgrune) [reply openlibrary/templates/books/breadcrumb_select.html]: Moved to mybooks-details.less!

File: static/css/components/mybooks-menu.less [hunk 3]
```diff
@@ -162,7 +162,7 @@
     }
 
     ul li {
-      background: @grey-e7e7e7;
+      background: #ddd;
```
Comment (mekarpeles) [inline static/css/components/mybooks-menu.less]: Add to `colors.less` -- no direct hex codes

File: openlibrary/templates/books/mobile_carousel.html [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  $if author_names:
+    $ byline = _(' by %(name)s', name=', '.join(author_names))
+  $else:
+    $ byline = ''
+
+  $ modifier = ''
+  $ work_key = book.key if book.key.startswith('/work') else book.get('work_key')
+
+  $ last_book = ''
+  $if index==4:
```
Comment (mekarpeles) [inline openlibrary/templates/books/mobile_carousel.html:43]: 4 appears to be a magic number number (for someone doing a quick pass).

Either variable or some other more intuitive check?

**Review by mekarpeles — APPROVED** [review]
Minor feedback; LGTM.
Nice to have; remove inline css, move colors to colors.less, and fix discrepency between mobile & desktop for carousel length 4 v. 5 items, document magic numbers

**Comment by cdrini** [comment]
Note: I noticed this caused a regression on the desktop my books page. The "Return ebook" button is missing! It should look like this: 
image

Something somewhere needs a `secondary_action=True`, probably.

File: openlibrary/templates/account/mybooks.html [hunk 5]
```diff
@@ -1,73 +1,229 @@
...
+          </div>
+        </li>
+        $if owners_page and ctx.user and ctx.user.in_sponsorship_beta():
+          <li>
+            <div class="carousel-section-header">
+              <a href="/people/$username/books/sponsorships" data-ol-link-track="MyBooksSidebar|Sponsorships" $('class=selected' if key == 'sponsorships' else '')>
+                $_('Sponsorships')
+                $ sponsorship_count = ''
+                $if counts.get('sponsorships', None):
+                  $ sponsorship_count = "%d(" % counts['sponsorships'] % ")"
```
Comment (mekarpeles) [inline openlibrary/templates/account/mybooks.html:220]: Somehow this got reverted -- should be
```suggestion
                  $ sponsorship_count = "(%d)" % counts.get('sponsorships')
```

File: openlibrary/templates/books/breadcrumb_select.html [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                    window.location = nav.value;
+                })
+
+            nav.addEventListener('keydown', ({ key }) => {
+                if (preventedKeys.has(key))
+                ignoreChange = true
+                else if (allowedKeys.has(key))
+                ignoreChange = false
+            })
+            })
```
Comment (jimchamp) [inline openlibrary/templates/books/breadcrumb_select.html:67]: @szgrune, can you explain the desired behavior here?  The selected breadcrumb wraps in FF, but does not in Chrome.  Also, what it the `' '` key, and how do I test it?

**Pull request merged**